src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/HeaderModelBinder.cs

```diff
@@ -21,50 +22,123 @@ public class HeaderModelBinder : IModelBinder
 
-            _logger.AttemptingToBindModel(bindingContext);
-
+            // Do not set ModelBindingResult to Failed on not finding the value in the header as we want the inner 
+            // modelbinder to do that. This would give a chance to the inner binder to add more useful information.
+            // For example, SimpleTypeModelBinder adds a model error when binding to let's say an integer and the
+            // model is null.
+            var request = bindingContext.HttpContext.Request;
             if (!request.Headers.ContainsKey(headerName))
             {
                 _logger.FoundNoValueInRequest(bindingContext);
             }
 
+            if (InnerModelBinder == null)
+            {
+                BindWithoutInnerBinder(bindingContext);
+                return;
+            }
+
+            var headerValueProvider = GetHeaderValueProvider(headerName, bindingContext);
+
+            // Capture the top level object here as entering nested scope would make it 'false' always.
+            var isTopLevelObject = bindingContext.IsTopLevelObject;
+
+            // Create a new binding scope in order to supply the HeaderValueProvider so that the binders like
+            // SimpleTypeModelBinder can find values from header.
+            ModelBindingResult result;
+            using (bindingContext.EnterNestedScope(
+                    bindingContext.ModelMetadata,
+                    fieldName: bindingContext.FieldName,
+                    modelName: bindingContext.ModelName,
+                    model: bindingContext.Model))
+            {
+                bindingContext.IsTopLevelObject = isTopLevelObject;
+                bindingContext.ValueProvider = headerValueProvider;
+
+                await InnerModelBinder.BindModelAsync(bindingContext);
+                result = bindingContext.Result;
+            }
+
+            bindingContext.Result = result;
+
+            _logger.DoneAttemptingToBindModel(bindingContext);
+        }
+
+        private HeaderValueProvider GetHeaderValueProvider(string headerName, ModelBindingContext bindingContext)
+        {
+            var request = bindingContext.HttpContext.Request;
+
+            // Prevent breaking existing users in scenarios where they are binding to a 'string' property
+            // and expect the whole comma separated string, if any, as a single string and not as a string array.
+            var values = Array.Empty<string>();
+            if (request.Headers.ContainsKey(headerName))
+            {
+                if (bindingContext.ModelMetadata.IsEnumerableType)
+                {
+                    values = request.Headers.GetCommaSeparatedValues(headerName);
+                }
+                else
+                {
+                    values = new[] { (string)request.Headers[headerName] };
+                }
+            }
+
+            return new HeaderValueProvider(CultureInfo.InvariantCulture, values);
```

Comment:
Don't pass in the culture or store it in a field. Use `CultureInfo.InvariantCulture` when creating the `ValueProviderResult`.